### ISSUE TITLE
Why no translateTo function?

### ISSUE THREAD

**Issue by magjac** [d3/d3-zoom#94]
There is scaleBy and scaleTo, but only translateBy. Is there a specific reason or is it just not implemented?

**Comment by mbostock** [comment]
Yes. It is not clear whether the coordinates *x*,*y* you pass to *zoom*.translateTo should depend on the transform’s scale. You have to decide whether you want this:

```js
zoom.translate(selection, d3.zoomIdentity.scale(k).translate(x, y));
```

Or this:

```js
zoom.translate(selection, d3.zoomIdentity.translate(x, y).scale(k));
```

Given the ambiguity, it seemed better to not implement this method, and use one of the two above instead.

**Comment by pmalouin** [comment]
Something else that I think would be nice is that zoom.translateTo could respect the translateExtent values automatically. Otherwise, it must be done prior to calling zoom.translate.

My use case is that I want to programmatically center the visualization on an element, but only if it would respect the translateExtent (i.e. it would not be respected if the element is on the edge of the visualization).

**Comment by baygeldin** [comment]
I'm aware of the ambiguity, but the reason I was looking for this method is exactly that I wanted a way to translate to a specific point with respect to `translateExtent`. And since other convinience methods like `translateBy` and `scaleTo` works like that I figured that such a method would be super handy. So I came up with a following wrapper for zoom generator (it's written in CoffeeScript, but it's easy to translate it to pure JS):

```coffeescript
zoom = () ->
    zoom = d3.zoom()

    # Custom "zoom.transform" that enforces translate and scale extents.
    zoom.strictTransform = (selection, transform) ->
      [[x0, y0], [x1, y1]] = zoom.translateExtent()
      [k0, k1] = zoom.scaleExtent()

      constrain = (transform, extent) ->
        dx0 = transform.invertX(extent[0][0]) - x0
        dx1 = transform.invertX(extent[1][0]) - x1
        dy0 = transform.invertY(extent[0][1]) - y0
        dy1 = transform.invertY(extent[1][1]) - y1

        x = if dx1 > dx0 then (dx0 + dx1) / 2 \
          else Math.min(0, dx0) or Math.max(0, dx1)
        y = if dy1 > dy0 then (dy0 + dy1) / 2 \
          else Math.min(0, dy0) or Math.max(0, dy1)

        transform.translate(x, y)

      selection.call(zoom.transform, () ->
        { x, y, k } = if typeof transform is 'function' then \
          transform.apply(this, arguments) else transform
        k = Math.max(k0, Math.min(k1, k))
        constrain(d3.zoomIdentity.translate(x, y).scale(k),
          zoom.extent().apply(this, arguments)))

    # Translates to position.
    zoom.translateTo = (selection, position) ->
      selection.call(zoom.strictTransform, () ->
        { x, y, k } = Object.assign({},
          d3.zoomTransform(this), position)
        d3.zoomIdentity.translate(x, y).scale(k))

    zoom
```

The `constrain` function is from `d3-zoom` sources. I'm not sure if this solution is completely correct (especially I'm not sure about transitions), but it works well for me. Anyway, I believe that either a strict version of `transform` or two different versions of `translateTo` (to deal with ambiguity) would be nice to have in D3.

**Comment by mbostock** [comment]
On second thought, I don’t think it is ambiguous because the *x* and *y* you pass to *zoom*.translateTo should be consistent with the *x* and *y* you pass to *zoom*.translateBy. That means that *t<sub>x1</sub>* = *k* × *x* and *t<sub>y1</sub>* = *k* × *y*, or in other words:

```js
zoom.translate(selection, d3.zoomIdentity.scale(k).translate(x, y));
```

I haven’t tested it, but there’s an implementation in #109. Is that what you want?

**Comment by baygeldin** [comment]
This seems like exactly what I needed. It'd be nice to see this feature in future releases. Thank you for such a quick response, this is great! :)

**Comment by mbostock** [comment]
I would love to see an example of why this is useful if you have time!

**Comment by baygeldin** [comment]
Sure, I came up with [a simple example that shows the problem](https://bl.ocks.org/baygeldin/16c9a7c23952ff2ade6c7708b88aa90f). My original use case is different and more complex (it also involves a custom brush, transitions, etc.), but to the number of reasons I can't provide it, sorry. I hope this helps.

**Comment by pmalouin** [comment]
Just pinpointing that @baygeldin's example uses the other translation order:
```
zoom.translate(selection, d3.zoomIdentity.translate(x, y).scale(k));
```
Wouldn't it be this form that is consistent with zoom.translateBy?

**Comment by mbostock** [comment]
@pmalouin No. That produces {k: k, x: x, y: y}, while the other order produces {k: k, x: k * x, y: k * y}, which is consistent with *zoom*.translateBy.

**Comment by baygeldin** [comment]
@pmalouin, seems like I made a mistake in a hurry. I changed the order and updated the example.

**Comment by mbostock** [comment]
The issue I see with the proposed implementation in #109 and @baygeldin’s use case is that *zoom*.translateTo would not be especially helpful as this method puts the specified point ⟨*x*, *y*⟩ at the origin ⟨0,0⟩. What you typically want instead is for the specified point to be at the center of the viewport.

Fortunately, there’s already logic in place to compute the [center of the extent](https://github.com/d3/d3-zoom/blob/v1.3.1/src/zoom.js#L139-L141), which is used in *zoom*.scaleTo and transitions. So I think we can tweak the proposed implementation accordingly, and then *zoom*.translateTo will be useful in any situation where you want to center something at the current zoom level (such as the [zoom transitions example](https://bl.ocks.org/mbostock/b783fbb2e673561d214e09c7fb5cedee)).

**Comment by baygeldin** [comment]
In my opinion, the proposed implementation is already quite useful, because it helps to avoid a lot of boilerplate with writing your own constrain logic or figuring out how to acomplish the same thing with `translateBy`. In fact in my original use case I do translate points to ⟨0,0⟩, but I just thought that for the example it'd seem nicer to translate them to the center of the viewport. But if there would be something that also helps with centering, it'd be nice.

**Comment by mbostock** [comment]
Okay, I tested the new #109 on the zoom transitions example, and it worked great! Will release soon as 1.4.0. Thanks for the motivation. 😁

**Comment by Pavnii** [comment]
I am using d3 v4 but still unable to use zoom.translateTo functionality.
